Initial internal support for multiple webviews

wusyong opened the pull request:
This patch improves support for multiple top-level browsing contexts aka “webviews” by adding several messages that allow the embedder to show, hide, move, and resize webviews in the compositor, and notify the embedder when painting order and focus changes.

This is a rebased version of #30648, while addressing review feedback given in #30767, #30840, #30841, and #30842. Notably, “browsers” have been renamed to “webviews”.

There are no user-facing changes in servoshell yet, but these changes can be used to create a multiple-document interface (#30785) or implement tabbed browsing (#31545). For the full design of this feature, see #30648.

## limitations
- desktop-style page zoom is still global
- browsers with transparent or translucent content will misbehave
- sending ShowWebView and HideWebView for one webview at a time can be inefficient
- if too many webviews are shown at once, the compositor can get stuck in a laggy state with pending epoch warnings (to reproduce, try [#31545](https://github.com/servo/servo/pull/31545) with [4.html](https://bucket.daz.cat/work/igalia/servo/4.html) and click “_blank” six or seven times)

## compositor
- WebViewManager (new) manages webview-related storage and painting order
    - it allows the compositor to store arbitrary data for each webview id
    - it keeps track of which webviews are visible, and what order to paint them in
- we now use the iframe-in-dummy-root-pipeline technique to composite the webviews together, not just when we have a non-unity pinch zoom

## servoshell
- WebViewManager (was WebView) keeps a copy of the webview rects and focus
- but for now, we always focus and show the newest webview only

## messages
embedder → compositor (no ipc)
- WindowResize (was Resize) no longer affects the viewport of individual webviews
- Zoom (existing) no longer affects the viewport of individual webviews
- ResetZoom (existing) no longer affects the viewport of individual webviews
- MoveResizeWebView (new) updates a webview’s rect, and sends new viewport to script in descendant pipelines
- ShowWebView (new) adds a webview to painting order, and notifies embedder of the new painting order
- HideWebView (new) removes a webview from painting order, and notifies embedder of the new painting order
- RaiseWebViewToTop (new) moves a webview to the top of the painting order, and notifies embedder accordingly

embedder →\* constellation
- SetWebViewThrottled (was WindowVisibility) has been reworked to be less confusing ([#31815](https://github.com/servo/servo/pull/31815), [#31816](https://github.com/servo/servo/pull/31816))

constellation → compositor
- CreateOrUpdateWebView (was SetFrameTree) now adds or updates a webview without clobbering others
- RemoveWebView (new) removes a webview from .webviews and .pipeline_details
- MoveResizeWebView (new) updates a webview’s rect, and sends new viewport to script in descendant pipelines
- ShowWebView (new) adds a webview to painting order, and notifies embedder of the new painting order
- HideWebView (new) removes a webview from painting order, and notifies embedder of the new painting order
- RaiseWebViewToTop (new) moves a webview to the top of the painting order, and notifies embedder accordingly

embedder ← constellation ←\* compositor
- ReadyToPresent (existing) now takes a vec of the ids of all visible webviews in painting order

\* all of these are in the same enum ConstellationMsg (components/shared/compositing/constellation_msg.rs), which is a very confusing design that conflates the embedder with the compositor

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes are related to #30593

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

wusyong committed:
Add multiple concurrent top-level browsing contexts

Co-authored-by: Delan Azabani <dazabani@igalia.com>

wusyong committed:
Rename variables and comments

There are some variable and comments still use browser as names.
This commit renames them to webview.

wusyong commented:
@delan I think the PR is ready to review now. Could you review it if you have time?
The last thing I'm not sure is `WebViewVisibilityChanged`. Should it be removed in favor of `ShowWebView` and `HideWebView` variants?

delan commented:
> The last thing I'm not sure is `WebViewVisibilityChanged`. Should it be removed in favor of `ShowWebView` and `HideWebView` variants?

WebViewVisibilityChanged notifies script that a webview has become invisible due to the native window or app activity, whereas ShowWebView and HideWebView tell the compositor to add or remove it from webrender (and notifies script of the new visibility, taking native window visibility into account). I think we should keep it for now, but I’m open to changing the model if something else makes more sense, like @mrobinson did in #30842 (9e307cfcefed26ecb65399ce12a7e2a7ef38e5d7).

delan commented on components/constellation/webview.rs:
```diff
@@ -93,6 +101,38 @@ impl<WebView> WebViewManager<WebView> {
         self.is_focused = false;
     }
+
+    pub fn set_webview_visibility(
+        &mut self,
+        top_level_browsing_context_id: TopLevelBrowsingContextId,
+        visible: bool,
+    ) {
+        debug_assert!(self.webviews.contains_key(&top_level_browsing_context_id));
+        if visible {
+            self.visible_webviews.insert(top_level_browsing_context_id);
+        } else {
+            self.visible_webviews.remove(&top_level_browsing_context_id);
+        }
+    }
+
+    pub fn set_native_window_visibility(&mut self, visible: bool) {
+        self.native_window_is_visible = visible;
+    }
```
given that native window visibility was removed (<https://github.com/servo/servo/pull/30842/commits/9e307cfcefed26ecb65399ce12a7e2a7ef38e5d7>), i think we want something like

```rust
fn mark_webview_shown(webview_id: TopLevelBrowsingContextId);
fn mark_webview_not_shown(webview_id: TopLevelBrowsingContextId);
fn mark_webview_invisible(webview_id: TopLevelBrowsingContextId);
fn mark_webview_not_invisible(webview_id: TopLevelBrowsingContextId);
```

wusyong replied:
Added in [`6d54829`](https://github.com/servo/servo/pull/31417/commits/6d5482974e6a17e98dd146626fe61a88aa582aee)

delan commented on components/compositing/compositor.rs:
```diff
@@ -974,53 +1072,165 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
-            frame_tree.pipeline.id
-        );
+    fn update_webview(&mut self, frame_tree: &SendableFrameTree) {
+        debug!("{}: Updating web view", frame_tree.pipeline.id);
```
nit: in logging and comments, use “webview” (one word)

wusyong replied:
fixed in ([`070398d`](https://github.com/servo/servo/pull/31417/commits/070398d0030a23be7ae9581d8a486ad473678ca6))

delan replied:
thanks! there are still 17 other instances of “web view” in this patch, can you update those too?

wusyong replied:
fixed in ([`ade8c27`](https://github.com/servo/servo/pull/31417/commits/ade8c27dfad50361a3c0d04567968cb297edc454))

delan reviewed (commented):
Thanks for rebasing this, it’s coming along well. I’ve edited in a description of the changes in this patch, and updated the description in #30648.

I guess we either need to finish and enable the multiview feature in this patch, or fix the viewport coordinates when multiview is disabled, but given how old this patch is getting I think we should do the latter.

wusyong committed:
Update log message from web view to webview

wusyong committed:
Revert offscreen_framebuffer_id rename

wusyong commented on components/constellation/webview.rs:
```diff
@@ -17,6 +17,12 @@ pub struct WebViewManager<WebView> {
 
     /// Whether the latest webview in focus order is currently focused.
     is_focused: bool,
+
+    /// The web view that would be visible in a containing native window.
+    visible_webviews: HashSet<TopLevelBrowsingContextId>,
+
+    /// Whether our native window is visible, or true if there is no such window.
+    native_window_is_visible: bool,
 }
```
I would like to know if we could simplify it with the set of `visible_webviews` only. This is also related to the comment above about `WebViewVisibilityChanged`. Could we keep only either `WebViewVisibilityChanged` or `ShowWebView`/`HideWebView` variant?

Because users might not understand internal parts like constellation and compositor, when I see the "visibility" configuration, I would expect it to be a boolean option to toggle as well. Could compositor and constellation share only one visibility set and reflect based on it?

wusyong replied:
After some discussion, we decide to replace with `shown_webviews` and `invisible_webviews`. We'll think about how to simplify it once multi windows support is here.

wusyong commented on components/servo/lib.rs:
```diff
@@ -883,13 +913,9 @@ where
     /// Return the OpenGL framebuffer name of the most-recently-completed frame when compositing to
     /// [`CompositeTarget::Fbo`], or None otherwise.
-    pub fn offscreen_framebuffer_id(&self) -> Option<u32> {
+    pub fn output_framebuffer_id(&self) -> Option<gl::GLuint> {
```
fixed in ([`41d6eea`](https://github.com/servo/servo/pull/31417/commits/41d6eea98a42ec7307a7d227e2e61f4ff37eac8e))

wusyong committed:
Rename all web view to webview

wusyong committed:
Cargo fmt

delan committed:
Fix viewport/event/clear coordinates when multiview is disabled

delan committed:
Only deprecate things when multiview is enabled

delan commented:
While testing these changes, I thought I found another regression with multiview disabled, but it actually affects main too (#31539), so I opened #31541 to make it easier to play with our multiview work.

wusyong committed:
Update WebViewManger with shown and invisible sets

Replace visible_webviews and native_window_is_visible with shown_webviews
and invisible_webviews. Add 4 more methods to set them accordingly. The
behavior of is_effectively_visible will return true if the wbview is in
shown_webviews set but not in invisible_webviews.

wusyong commented on components/constellation/constellation.rs:
```diff
@@ -1543,7 +1607,16 @@ where
+                        .webviews
+                        .is_effectively_visible(top_level_browsing_context_id);
+                    self.notify_webview_visibility(top_level_browsing_context_id, visible);
+                }
```
I update the behavior of `RaiseWebViewToTop`, `ShowWebView`, `HideWebView`, and `WebViewVisibilityChanged`. Let me know if those are correct.
I haven't renamed it to `MarkWebViewInvisible` because I think there's still need a way to `mark_webview_no_invisible`.

delan replied:
looks good, except for a few issues:

- HideWebView should mark_webview_not_shown, not mark_webview_invisible
- WebViewVisibilityChanged should not call mark_webview_shown or mark_webview_not_shown
- WebViewManager.shown_webviews doc comment should be “webviews that are being shown by the embedder, regardless of whether they are actually visible”
- WebViewManager.invisible_webviews doc comment should be “webviews that are known to be invisible, regardless of whether they are being shown”

wusyong replied:
Updated in [`8083275`](https://github.com/servo/servo/pull/31417/commits/8083275411114606e24954af39e7876d287e89c6), and the variant is renamed in [`258d520`](https://github.com/servo/servo/pull/31417/commits/258d5202ea13ab2c2e2dee03a810c061f2e22b49)

wusyong committed:
Update variant behaviors

wusyong committed:
Rename WebViewVisibilityChanged to MarkWebViewInvisible

wusyong committed:
Fix unit test by marking id 3 visible again

wusyong committed:
Update MarkWebViewInvisible and add UnmarkWebViewInvisible

delan committed:
Improve logging of webview-related embedder events

delan commented on components/shared/compositing/lib.rs:
```diff
@@ -106,8 +106,18 @@ pub enum CompositorMsg {
+    UpdateWebView(SendableFrameTree),
+    /// Remove a webview.
+    RemoveWebView(TopLevelBrowsingContextId),
+    /// Make a webview visible.
```
```suggestion
    /// Move and/or resize a webview to the given rect.
```

delan replied:
done in de1d967fba28beef0dd36bac009ea09cd4250d50

delan commented on components/shared/compositing/lib.rs:
```diff
@@ -106,8 +106,18 @@ pub enum CompositorMsg {
+    RemoveWebView(TopLevelBrowsingContextId),
+    /// Make a webview visible.
+    MoveResizeWebView(TopLevelBrowsingContextId, DeviceRect),
+    /// Make a webview visible.
```
```suggestion
    /// Start painting a webview.
```

delan replied:
done in de1d967fba28beef0dd36bac009ea09cd4250d50

delan commented on components/shared/compositing/lib.rs:
```diff
@@ -106,8 +106,18 @@ pub enum CompositorMsg {
+    MoveResizeWebView(TopLevelBrowsingContextId, DeviceRect),
+    /// Make a webview visible.
+    ShowWebView(TopLevelBrowsingContextId),
+    /// Make a webview invisible.
```
```suggestion
    /// Stop painting a webview.```

delan replied:
done in de1d967fba28beef0dd36bac009ea09cd4250d50

delan commented on components/shared/compositing/lib.rs:
```diff
@@ -106,8 +106,18 @@ pub enum CompositorMsg {
+    ShowWebView(TopLevelBrowsingContextId),
+    /// Make a webview invisible.
+    HideWebView(TopLevelBrowsingContextId),
+    /// Make a webview visible and paint on top of all others.
```
```suggestion
    /// Start painting a webview on top of all others.
```

delan replied:
done in de1d967fba28beef0dd36bac009ea09cd4250d50

delan commented on components/shared/compositing/constellation_msg.rs:
```diff
@@ -60,6 +61,14 @@ pub enum ConstellationMsg {
     CloseWebView(TopLevelBrowsingContextId),
     /// Panic a top level browsing context.
     SendError(Option<TopLevelBrowsingContextId>, String),
+    /// Make a top-level browsing context visible.
```
```suggestion
    /// Move and/or resize a webview to the given rect.
```

delan replied:
done in de1d967fba28beef0dd36bac009ea09cd4250d50

delan commented on components/shared/compositing/constellation_msg.rs:
```diff
@@ -60,6 +61,14 @@ pub enum ConstellationMsg {
     SendError(Option<TopLevelBrowsingContextId>, String),
+    /// Make a top-level browsing context visible.
+    MoveResizeWebView(TopLevelBrowsingContextId, DeviceRect),
+    /// Make a top-level browsing context visible.
```
```suggestion
    /// Start painting a webview.
```

delan replied:
done in de1d967fba28beef0dd36bac009ea09cd4250d50

delan commented on components/shared/compositing/constellation_msg.rs:
```diff
@@ -60,6 +61,14 @@ pub enum ConstellationMsg {
+    MoveResizeWebView(TopLevelBrowsingContextId, DeviceRect),
+    /// Make a top-level browsing context visible.
+    ShowWebView(TopLevelBrowsingContextId),
+    /// Make a top-level browsing context invisible.
```
```suggestion
    /// Stop painting a webview.
```

delan replied:
done in de1d967fba28beef0dd36bac009ea09cd4250d50

delan commented on components/shared/compositing/constellation_msg.rs:
```diff
@@ -60,6 +61,14 @@ pub enum ConstellationMsg {
+    ShowWebView(TopLevelBrowsingContextId),
+    /// Make a top-level browsing context invisible.
+    HideWebView(TopLevelBrowsingContextId),
+    /// Make a top-level browsing context visible and paint on top of all others.
```
```suggestion
    /// Start painting a webview on top of all others.
```

delan commented on components/shared/compositing/constellation_msg.rs:
```diff
@@ -76,14 +85,18 @@ pub enum ConstellationMsg {
     MediaSessionAction(MediaSessionActionType),
-    /// The visibility of the webview has changed.
-    WebViewVisibilityChanged(TopLevelBrowsingContextId, bool),
+    /// Mark webview to be invisible, regardless of whether it is being shown
```
```suggestion
    /// Mark webview as invisible due to external factors, regardless of whether it is being painted.
```

delan replied:
done in de1d967fba28beef0dd36bac009ea09cd4250d50

delan commented on components/shared/compositing/constellation_msg.rs:
```diff
@@ -76,14 +85,18 @@ pub enum ConstellationMsg {
-    WebViewVisibilityChanged(TopLevelBrowsingContextId, bool),
+    /// Mark webview to be invisible, regardless of whether it is being shown
+    MarkWebViewInvisible(TopLevelBrowsingContextId),
+    /// Unmark webview to be invisible, it might not be visible if it's not shown
```
```suggestion
    /// Unmark webview as invisible due to external factors. The webview may remain invisible if it's not being painted.
```

delan replied:
done in de1d967fba28beef0dd36bac009ea09cd4250d50

delan commented on components/shared/embedder/lib.rs:
```diff
@@ -262,6 +264,7 @@ impl Debug for EmbedderMsg {
             EmbedderMsg::WebViewClosed(..) => write!(f, "WebViewClosed"),
             EmbedderMsg::WebViewFocused(..) => write!(f, "WebViewFocused"),
             EmbedderMsg::WebViewBlurred => write!(f, "WebViewUnfocused"),
```
```suggestion
            EmbedderMsg::WebViewBlurred => write!(f, "WebViewBlurred"),
```

delan replied:
done in de1d967fba28beef0dd36bac009ea09cd4250d50

delan reviewed (commented):
Great updates, this is almost ready!

wusyong commented on components/script/dom/bindings/buffer_source.rs:
Sorry, it's not. I guess it's because I use different version of nightly to do formatting.

delan replied:
fixed in de1d967fba28beef0dd36bac009ea09cd4250d50

wusyong committed:
Update format and doc comments

delan commented on ports/servoshell/app.rs:
```diff
@@ -284,7 +277,8 @@ impl App {
             }
-            if !consumed {
+            // TODO decide how to prioritise events between egui and servo
+            if true || !consumed {
```
done in de1d967fba28beef0dd36bac009ea09cd4250d50

delan commented on components/constellation/webview.rs:
```diff
@@ -185,6 +236,27 @@ mod test {
+
+        webviews.mark_webview_not_invisible(top_level_id(0, 1));
+        webviews.mark_webview_not_invisible(top_level_id(0, 3));
+        webviews.mark_webview_not_shown(top_level_id(0, 1));
```
done in de1d967fba28beef0dd36bac009ea09cd4250d50

delan committed:
Merge from main

delan committed:
Clean up doc comments

delan reviewed (approved):
Great work! I fixed merge conflicts in 60b0ef86a82e7fdfc0dba73e04548130cd6dfbae, and cleaned up and synced some outdated doc comments in 15f7dc88a293184d97a69efb09c446e4c7ced99d.

@mrobinson, would you like a pass at this?

delan committed:
Use webview_id and WebViewId in webview-related code

delan committed:
Allow webview Show and Raise events to optionally hide all others

mrobinson commented on components/compositing/compositor.rs:
```diff
@@ -1065,28 +1276,47 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
     }
 
+    #[cfg_attr(feature = "multiview", deprecated)]
     fn send_window_size(&mut self, size_type: WindowSizeType) {
-        let dppx = self.page_zoom * self.embedder_coordinates.hidpi_factor;
+        self.update_webrender_document_view();
```
If we do end up keeping both code paths, I think it makes sense to combine these two functions into something like `update_webrender_document_view_and_send_window_size()` to avoid the conditional compilation at each callsite.

delan replied:
removed in f3bb2c369948231facccf54ad7b8e45f1f8b0e8a

mrobinson commented on components/compositing/compositor.rs:
```diff
@@ -1152,9 +1396,10 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
-            Some(result) => result,
-            None => return,
+        let Some(result) = self.hit_test_at_device_point(point) else {
+            // TODO notify embedder?
+            // let msg = ConstellationMsg::UnconsumedEvent(...);
+            return;
```
I would remove the commented out code here.

In general, we should be notifying the embedder when a click isn't handled, but also when it hits a WebView, but ultimately isn't consumed as well, I think.

delan replied:
done in 4f22eec48373d3596c3c70ca7932fcdff91883ee

mrobinson commented on components/compositing/compositor.rs:
```diff
@@ -1847,6 +2102,10 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
                     // and check if it is the one layout is expecting,
                     let epoch = Epoch(epoch);
                     if *pending_epoch != epoch {
+                        warn!(
```
Is this really a warning case or is this something that we might expect in the normal execution of the program? If it's something normal, `debug!()` or `trace!()` is a better option.

delan replied:
i’m genuinely not sure, do you know? all i know for now is that it doesn’t really happen outside of some multiview-related fault scenarios, like failing to update the frame tree or having too many webviews open (see limitations), which makes me lean towards this shouldn’t happen.

mrobinson commented on components/compositing/compositor.rs:
```diff
@@ -1951,14 +2210,14 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
-        if let Err(e) = self.constellation_chan.send(msg) {
-            warn!("Sending event to constellation failed ({:?}).", e);
+        for (&top_level_browsing_context_id, _) in self.webviews.painting_order() {
+            let msg = ConstellationMsg::ReadyToPresent(top_level_browsing_context_id);
```
Can the `ReadyToPresent` carry an array now versus sending so many different messages?

delan replied:
done in 32aaf283ee1c6dc6d840544d8133dba402ecb552

mrobinson commented on components/compositing/compositor.rs:
```diff
@@ -2002,24 +2261,43 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
-        gl.clear(gleam::gl::COLOR_BUFFER_BIT);
-        gl.disable(gleam::gl::SCISSOR_TEST);
+
+        if cfg!(feature = "multiview") {
+            // Clear the viewport rect of each top-level browsing context.
+            for (_, webview) in self.webviews.painting_order() {
+                let rect = self
+                    .embedder_coordinates
+                    .flipped_rect(&webview.rect.to_i32());
+                gl.scissor(
+                    rect.origin.x,
+                    rect.origin.y,
+                    rect.size.width,
+                    rect.size.height,
+                );
+                gl.enable(gleam::gl::SCISSOR_TEST);
+                gl.clear(gleam::gl::COLOR_BUFFER_BIT);
+                gl.disable(gleam::gl::SCISSOR_TEST);
+            }
+        } else {
+            let viewport = self.embedder_coordinates.get_flipped_viewport();
+            gl.scissor(
+                viewport.origin.x,
+                viewport.origin.y,
+                viewport.size.width,
+                viewport.size.height,
+            );
+            gl.enable(gleam::gl::SCISSOR_TEST);
+            gl.clear(gleam::gl::COLOR_BUFFER_BIT);
+            gl.disable(gleam::gl::SCISSOR_TEST);
+        }
```
Hrm... So all of these views use the same GL context? I think that's an issue. They should be isolated in GPU memory from one another.

mrobinson replied:
After considering this for a while, I think this isn't as large of an issue as I thought. Each page is sending display lists to WebRender which could potentially be in a different process along with this compositor. This separation should be enough.

mrobinson replied:
After considering this for a while, I think this isn't as large of an issue as I thought. Each page is sending display lists to WebRender which could potentially be in a different process along with this compositor. This separation should be enough.

delan replied:
ok, no worries!

mrobinson commented on components/compositing/webview.rs:
```diff
@@ -0,0 +1,188 @@
+use msg::constellation_msg::TopLevelBrowsingContextId;
+
+#[derive(Debug)]
+pub struct WebViewManager<WebView> {
```
Why is this generic on WebView? Should there only be a single implementation of WebView?

delan replied:
since WebViewManager only uses the type for opaque storage, the unit tests can store dummy values that are cloneable and debuggable and so on, instead of all the data we actually store against each webview, making the tests easier to write and understand. same goes for WebViewManager in the constellation.

mrobinson commented on components/compositing/webview.rs:
```diff
@@ -0,0 +1,188 @@
+#[derive(Debug)]
+pub struct WebViewManager<WebView> {
+    /// Our top-level browsing contexts. In the WebRender scene, their pipelines are the children of
+    /// a single root pipeline that also applies any pinch zoom transformation.
```
So after the WebRender upgrade, the root pipeline will apply pinch zoom, page zoom, and also DPI scaling. :sweat_smile: In fact, if possible perhaps this can land after the upgrade, as it simplifies things in the compositor a bit.

```suggestion
    /// a single root pipeline that also applies any pinch zoom transformation.
```

delan replied:
sounds good! thanks for offering to rebase :)

mrobinson commented on components/compositing/webview.rs:
```diff
@@ -0,0 +1,188 @@
+    painting_order: Vec<TopLevelBrowsingContextId>,
+}
+
+impl<WebView> Default for WebViewManager<WebView> {
+    fn default() -> Self {
+        Self {
+            webviews: Default::default(),
+            painting_order: Default::default(),
+        }
+    }
+}
```
I think you can just derive from `Default` for this?

wusyong replied:
Fixed in [8c01639](https://github.com/servo/servo/pull/31417/commits/8c01639b1993c3b6c5fa57c1fcfc412716cf6f05)

mrobinson commented on components/compositing/windowing.rs:
```diff
@@ -103,8 +112,10 @@ pub enum EmbedderEvent {
-    /// The visibility of the webview has changed.
-    WebViewVisibilityChanged(TopLevelBrowsingContextId, bool),
+    /// Mark webview as invisible due to external factors, regardless of whether it is being painted.
+    MarkWebViewInvisible(TopLevelBrowsingContextId),
+    /// Unmark webview as invisible due to external factors. The webview may remain invisible if it's not being painted.
+    UnmarkWebViewInvisible(TopLevelBrowsingContextId),
```
Let's do a single message here `UpdateWebViewVisibility` with an enum argument.

delan replied:
the name UpdateWebViewVisibility may mislead consumers into believing that (true) and (false) will show and hide the webview, whereas what we’re actually doing is notifying the compositor that the webview is (or is no longer) invisible due to external factors. we tried to hint at this more passive and specific meaning with “Mark” and “Unmark”, but maybe there’s a better way?

mrobinson commented on components/compositing/windowing.rs:
```diff
@@ -211,15 +228,21 @@ pub struct EmbedderCoordinates {
-    pub fn get_flipped_viewport(&self) -> DeviceIntRect {
+    /// Flip the given rect.
+    /// This should be used when drawing directly to the framebuffer with OpenGL commands.
+    pub fn flipped_rect(&self, rect: &DeviceIntRect) -> DeviceIntRect {
```
```suggestion
    pub fn flip_rect(&self, rect: &DeviceIntRect) -> DeviceIntRect {
```

In order to make the first word of the function a verb.

wusyong replied:
Renamed to `flip_rect` in [8c01639](https://github.com/servo/servo/pull/31417/commits/8c01639b1993c3b6c5fa57c1fcfc412716cf6f05)

mrobinson commented on components/constellation/constellation.rs:
```diff
@@ -5396,24 +5472,25 @@ where
-        if let Some(focused_webview_id) = self.webviews.focused_webview().map(|(id, _)| id) {
-            if top_level_browsing_context_id != focused_webview_id {
-                return;
+    fn update_webview(&mut self, top_level_browsing_context_id: TopLevelBrowsingContextId) {
```
I think `send_frame_tree_if_focused()` is a bit more descriptive of what this function is actually doing. "Update" is so generic.

wusyong replied:
Renamed to `send_frame_tree_if_focused` in [8c01639](https://github.com/servo/servo/pull/31417/commits/8c01639b1993c3b6c5fa57c1fcfc412716cf6f05)

mrobinson commented on components/constellation/constellation.rs:
```diff
@@ -5396,24 +5472,25 @@ where
-            if top_level_browsing_context_id != focused_webview_id {
-                return;
+    fn update_webview(&mut self, top_level_browsing_context_id: TopLevelBrowsingContextId) {
+        if !cfg!(feature = "multiview") {
+            if let Some(focused_webview_id) = self.webviews.focused_webview().map(|(id, _)| id) {
+                if top_level_browsing_context_id != focused_webview_id {
+                    return;
+                }
+            } else {
+                self.webviews.focus(top_level_browsing_context_id);
             }
         }
```
I'm a bit confused as to why this only happens in the case that multiview is disabled. In any case, I think this should be using `any()`.

delan replied:
i agree that this is confusing, i think this was a combination of the way the patch evolved ([1](https://github.com/servo/servo/pull/30648/commits/6cd285f51644e3abc8b046dd17730d5762049644#diff-95aa2eb890fe794918afdadfb6a89e026ec80222ceb8e5a21768d528e9ce036c), [2](https://github.com/servo/servo/pull/30648/commits/510c6d4215d20b379d38ad9617fb20b658f65e2e#diff-95aa2eb890fe794918afdadfb6a89e026ec80222ceb8e5a21768d528e9ce036c)) and erring on the side of caution ([3](https://github.com/servo/servo/pull/30648/commits/f6403326bee652fa6cc84de75cb6bfa92f446c1f#diff-95aa2eb890fe794918afdadfb6a89e026ec80222ceb8e5a21768d528e9ce036c)), but then again the version we landed was unconditional ([4](https://github.com/servo/servo/pull/30842/commits/d3ce4a7280ea42555fab96385f568d42f8d6cb4d#diff-95aa2eb890fe794918afdadfb6a89e026ec80222ceb8e5a21768d528e9ce036cL5391-R5453)). i’ll look into it :)

delan replied:
305a50b9e632382b434faec61dce9abf7d0b6b01 removes both the feature gate and the focus check itself, hopefully the tests still pass!

mrobinson commented on components/constellation/webview.rs:
```diff
@@ -17,6 +17,12 @@ pub struct WebViewManager<WebView> {
     is_focused: bool,
+
+    /// Webviews that are being shown by the embedder, regardless of whether they are actually visible.
+    shown_webviews: HashSet<TopLevelBrowsingContextId>,
+
+    /// Webviews that are known to be invisible, regardless of whether they are being shown.
+    invisible_webviews: HashSet<TopLevelBrowsingContextId>,
```
The "shown" versus "visible" distinction is a bit hard to grasp! Is it something like "in the scene" versus "in the scene, but hidden?"

delan replied:
clarified in 9947d44e0c0b5e41128725e0e8e6cf8919890b60

mrobinson commented on components/servo/lib.rs:
```diff
@@ -217,6 +217,21 @@ pub struct InitializedServo<Window: WindowMethods + 'static + ?Sized> {
     pub browser_id: TopLevelBrowsingContextId,
 }
 
+/// Utility to help forward embedder event to constellation.
+macro_rules! forward_to_constellation {
+    ($self:ident, $variant:ident $(($($rest:tt),*))?) => {
+        {
+            let msg = ConstellationMsg::$variant$(($($rest),*))?;
+            if let Err(e) = $self.constellation_chan.send(msg) {
+                warn!(
+                    "Sending {} message to constellation failed ({:?}).",
+                    stringify!($variant), e,
+                );
+            }
+        }
+    };
+}
```
I think this will be easier to read as a function, if possible and also friendlier for language servers in the IDE. Just make sure that the message supports `Debug`.

delan replied:
done in 5364e71b8372de9e2c01ed3b8e5c81f59c93efe8

mrobinson commented on components/servo/lib.rs:
```diff
@@ -773,11 +803,21 @@ where
+            EmbedderEvent::MarkWebViewInvisible(webview_id) => {
+                let msg = ConstellationMsg::MarkWebViewInvisible(webview_id);
+                if let Err(e) = self.constellation_chan.send(msg) {
+                    warn!(
+                        "Sending MarkWebViewInvisible to constellation failed ({:?}).",
+                        e
+                    );
```
```suggestion
                    warn!("Sending MarkWebViewInvisible to constellation failed ({error:?}).");
```

wusyong replied:
Fixed in [8c01639](https://github.com/servo/servo/pull/31417/commits/8c01639b1993c3b6c5fa57c1fcfc412716cf6f05)

mrobinson commented on components/servo/lib.rs:
```diff
@@ -773,11 +803,21 @@ where
                 if let Err(e) = self.constellation_chan.send(msg) {
                     warn!(
-                        "Sending WebViewVisibilityChanged to constellation failed ({:?}).",
+                        "Sending UnmarkWebViewInvisible to constellation failed ({:?}).",
```
Sorry this suggestion won't apply cleanly due to a limitation of GitHub:

```suggestion
                   warn!("Sending UnmarkWebViewInvisible to constellation failed ({error:?}).");
```

wusyong replied:
Fixed in [8c01639](https://github.com/servo/servo/pull/31417/commits/8c01639b1993c3b6c5fa57c1fcfc412716cf6f05)

mrobinson commented on ports/servoshell/minibrowser.rs:
```diff
@@ -190,9 +194,28 @@ impl Minibrowser {
             CentralPanel::default()
                 .frame(Frame::none())
                 .show(ctx, |ui| {
+                    let Pos2 { x, y } = ui.cursor().min;
+                    let origin = Point2D::new(x, y);
+                    let Vec2 { x, y } = ui.available_size();
+                    let size = Size2D::new(x, y);
+                    let rect = Rect::new(origin, size) * scale;
```
```suggestion
                    let Pos2 { x, y } = ui.cursor().min;
                    let Vec2 { x, y } = ui.available_size();
                    let rect = Rect::new(
                        Point2D::new(x, y),
                        Size2D::new(width, height)
                     ) * scale;
```

wusyong replied:
Fixed in [8c01639](https://github.com/servo/servo/pull/31417/commits/8c01639b1993c3b6c5fa57c1fcfc412716cf6f05)

mrobinson commented on ports/servoshell/webview.rs:
```diff
@@ -542,14 +564,28 @@ where
+                    let toolbar = self.window.toolbar_height().get();
+
+                    // Adjust for our toolbar height.
+                    // TODO adjust for egui window decorations if we end up using those
```
```suggestion
                    // TODO: Adjust for egui window decorations if we end up using those.
```

wusyong replied:
Fixed in [8c01639](https://github.com/servo/servo/pull/31417/commits/8c01639b1993c3b6c5fa57c1fcfc412716cf6f05)

mrobinson commented on ports/servoshell/webview.rs:
```diff
@@ -564,6 +600,31 @@ where
+                        let mut newest_webview_is_visible = false;
+
+                        // Hide any visible browsers other than the most recently created.
+                        // TODO stop doing this once we have full multiple browser support
```
```suggestion
                        // TODO: Stop doing this once we have full multiple browser support.
```

wusyong replied:
Fixed in [8c01639](https://github.com/servo/servo/pull/31417/commits/8c01639b1993c3b6c5fa57c1fcfc412716cf6f05)

mrobinson commented on ports/servoshell/webview.rs:
```diff
@@ -564,6 +600,31 @@ where
+                        }
+
+                        // If the most recently created browser is not visible, show it.
+                        // TODO stop doing this once we have full multiple browser support
```
```suggestion
                        // TODO: Stop doing this once we have full multiple browser support.
```

wusyong replied:
Fixed in [8c01639](https://github.com/servo/servo/pull/31417/commits/8c01639b1993c3b6c5fa57c1fcfc412716cf6f05)

mrobinson reviewed (commented):
There's a lot of great stuff here.

I haven't had a good look at the changes to minibrowser yet, as I'm less familiar with that code, but hopefully this can get the conversation started.

wusyong committed:
Address style and naming changes

wusyong commented on components/compositing/compositor.rs:
```diff
@@ -1749,7 +2004,7 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
     ) -> Result<Option<Image>, UnableToComposite> {
         if self.waiting_on_present {
-            debug!("tried to composite while waiting on present");
+            trace!("tried to composite while waiting on present");
```
Fixed in [8c01639](https://github.com/servo/servo/pull/31417/commits/8c01639b1993c3b6c5fa57c1fcfc412716cf6f05)

wusyong commented on components/compositing/compositor.rs:
```diff
@@ -991,53 +1089,166 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
-            "Setting the frame tree for pipeline {:?}",
-            frame_tree.pipeline.id
-        );
+    fn update_webview(&mut self, frame_tree: &SendableFrameTree) {
```
Renamed to `set_frame_tree_for_webview` in [8c01639](https://github.com/servo/servo/pull/31417/commits/8c01639b1993c3b6c5fa57c1fcfc412716cf6f05)

wusyong commented on components/compositing/compositor.rs:
```diff
@@ -952,20 +1041,34 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
+        self.generate_frame(&mut txn);
+        self.webrender_api
+            .send_transaction(self.webrender_document, txn);
+    }
```
Fixed in [8c01639](https://github.com/servo/servo/pull/31417/commits/8c01639b1993c3b6c5fa57c1fcfc412716cf6f05)

wusyong commented on components/compositing/compositor.rs:
```diff
@@ -952,20 +1041,34 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
 
+    /// Set the root pipeline for our WebRender scene to a display list that consists of an iframe
+    /// for each visible top-level browsing context, applying the pinch zoom transformation if any.
+    fn update_root_pipeline_with_txn(&self, transaction: &mut Transaction) {
```
Fixed in [8c01639](https://github.com/servo/servo/pull/31417/commits/8c01639b1993c3b6c5fa57c1fcfc412716cf6f05)

wusyong commented on components/compositing/compositor.rs:
```diff
@@ -379,6 +388,12 @@ pub enum CompositeTarget {
+    pub pipeline_id: Option<PipelineId>,
+    pub rect: DeviceRect,
+}
+
```
Fixed in [8c01639](https://github.com/servo/servo/pull/31417/commits/8c01639b1993c3b6c5fa57c1fcfc412716cf6f05)

delan committed:
Rename UpdateWebView to UpdateFrameTreeForWebView

delan committed:
constellation: send frame tree unconditionally over focus and feature

delan committed:
Clarify shown and invisible sets in constellation WebViewManager

delan commented on components/compositing/compositor.rs:
```diff
@@ -1065,28 +1276,47 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
+        let children = self
+            .pipeline_details
+            .iter()
+            .filter(|(_, pipeline_details)| {
+                pipeline_details.parent_pipeline_id == Some(pipeline_id)
+            })
+            .map(|(&pipeline_id, _)| pipeline_id)
```
i agree, though i’m not so sure we should in this case. this feels confusing to me…

```rust
.filter_map(|(pipeline_id, pipeline_details)| {
    pipeline_details.parent_pipeline_id.filter(|id| id == pipeline_id))
})
```

…whereas this just feels a bit awkward. what do you think?

```rust
.filter_map(|(pipeline_id, pipeline_details)| {
    if pipeline_details.parent_pipeline_id == Some(pipeline_id) {
        Some(pipeline_id)
    } else {
        None
    }
})
```

delan committed:
Eliminate forward_to_constellation!()

delan committed:
Actually remove the unused macro

delan commented on components/compositing/compositor.rs:
```diff
@@ -1198,12 +1443,14 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
-        };
-        if self.pipeline(root_pipeline_id).is_none() {
-            return vec![];
+        if !cfg!(feature = "multiview") {
+            let root_pipeline_id = match self.root_content_pipeline.id {
+                Some(root_pipeline_id) => root_pipeline_id,
+                None => return vec![],
+            };
+            if self.pipeline(root_pipeline_id).is_none() {
+                return vec![];
+            }
```
to be honest, i don’t know why we had this check in the first place. maybe we can remove it…

delan replied:
done in f3bb2c369948231facccf54ad7b8e45f1f8b0e8a; hopefully the tests pass :)

delan commented on components/compositing/compositor.rs:
```diff
@@ -991,53 +1089,166 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
+            let viewport_size = LayoutSize::from_untyped(viewport_size.to_untyped());
+            for (_, webview) in self.webviews.painting_order() {
+                if let Some(pipeline_id) = webview.pipeline_id {
+                    let rect = webview.rect / dppx;
+                    builder.push_iframe(
+                        LayoutRect::from_untyped(&rect.to_untyped()),
+                        LayoutRect::from_untyped(&rect.to_untyped()),
+                        &SpaceAndClipInfo {
+                            spatial_id: zoom_reference_frame,
+                            clip_id: ClipId::root(pipeline_id.to_webrender()),
+                        },
+                        pipeline_id.to_webrender(),
+                        true,
+                    );
+                }
+            }
+            viewport_size
+        } else {
+            let root_content_pipeline = self
+                .root_content_pipeline
+                .id
+                .expect("checked above")
+                .to_webrender();
+            let viewport_size = LayoutSize::new(
+                self.embedder_coordinates.get_viewport().width() as f32,
+                self.embedder_coordinates.get_viewport().height() as f32,
+            );
+            let viewport_rect = LayoutRect::new(LayoutPoint::zero(), viewport_size);
+            builder.push_iframe(
+                viewport_rect,
+                viewport_rect,
+                &SpaceAndClipInfo {
+                    spatial_id: zoom_reference_frame,
+                    clip_id: ClipId::root(root_pipeline),
+                },
+                root_content_pipeline,
+                true,
+            );
+            viewport_size
+        };
```
hmm, good point. while there are a bunch of bugs to iron out with multiple webviews, i don’t think any of them are actually regressions, and the feature does work well with a single webview. i’ll see what i can do :)

delan replied:
done in f3bb2c369948231facccf54ad7b8e45f1f8b0e8a; hopefully the tests pass :)

delan commented on components/compositing/compositor.rs:
```diff
@@ -991,53 +1089,166 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
+        }
 
-        self.create_pipeline_details_for_frame_tree(&frame_tree);
+        if !cfg!(feature = "multiview") {
+            let webview_ids = self
+                .webviews
+                .painting_order()
+                .map(|(&id, _)| id)
+                .collect::<Vec<_>>();
+            for webview_id in webview_ids {
+                if webview_id != top_level_browsing_context_id {
+                    self.webviews.hide(webview_id);
+                }
+            }
+            self.webviews.raise_to_top(top_level_browsing_context_id);
+        }
```
yeah, SetFrameTree (now CreateOrUpdateWebView) used to imply showing the webview and hiding all others, so this updates the WebViewManager to keep it consistent with that.

delan committed:
Don’t gate compositor changes on multiview feature flag

delan commented on components/compositing/compositor.rs:
```diff
@@ -952,20 +1041,34 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
+                    .to_webrender();
+                transaction.set_root_pipeline(root_content_pipeline);
+                return;
+            }
```
done in f3bb2c369948231facccf54ad7b8e45f1f8b0e8a

delan committed:
Update todo in mouse event dispatch

delan committed:
Pass all visible webview ids in a single ReadyToPresent message

mrobinson committed:
Merge branch 'main' into multi-tlbc

delan committed:
Fix compile and lint errors

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
Merge from main

delan committed:
servoshell: fix gap between minibrowser toolbar and webview

delan committed:
Merge from #31774

delan committed:
Fix failure in /_mozilla/mozilla/window_resizeTo.html

delan committed:
Merge from main

delan committed:
Fix compile warnings

delan commented:
Fixed the test failure, but there seems to be a regression after merging in the WebRender update (1ae4cd216d0fec96f8b0ebefbb48137187ad09d5). When clicking on a page scrolled down to some non-zero y, the page briefly flickers back to the top.

mrobinson commented:
> Fixed the test failure, but there seems to be a regression after merging in the WebRender update ([1ae4cd2](https://github.com/servo/servo/commit/1ae4cd216d0fec96f8b0ebefbb48137187ad09d5)). When clicking on a page scrolled down to some non-zero y, the page briefly flickers back to the top.

This might be due to us sending the wrong scroll coordinates to WebRender. I'll take a look at this now.

delan committed:
Remove stray dbg!()

delan commented:
> This might be due to us sending the wrong scroll coordinates to WebRender. I'll take a look at this now.

Tracking this in #31807.

delan committed:
Merge from main

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
Remove confusing “effectively visible” logic (see #31815, #31816)

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
Allow embedder to show/hide/raise webviews without ipc

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
Update root pipeline only when painting order actually changes

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
Stop gating old focus and SetFrameTree behaviour behind Cargo feature

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
Don’t do anything in response to WebViewPaintingOrder

delan committed:
Remove WebViewPaintingOrder, since its payload is unreliable

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan commented:
@mrobinson, this patch should be ready for another review. I made the changes we discussed…

- [Remove confusing “effectively visible” logic (see #31815, #31816)](https://github.com/servo/servo/pull/31417/commits/0ad18c13ef0c3543c88b4ddc3379dc0ea45a470f)
  - [Allow embedder to show/hide/raise webviews without ipc](https://github.com/servo/servo/pull/31417/commits/371affc746730e5bb369c3710e3132151e579d4c)
- [Update root pipeline only when painting order actually changes](https://github.com/servo/servo/pull/31417/commits/47ccd1cfc5ee45e9236170a0a38d7b9b6d401645)
- [Stop gating old focus and SetFrameTree behaviour behind Cargo feature](https://github.com/servo/servo/pull/31417/commits/adcb4627f0c585fd3e8f15af623061a5a51280bb)

…plus a few other changes to fix test failures.

- [Allow webview Show and Raise events to optionally hide all others](https://github.com/servo/servo/pull/31417/commits/e055056f45af4991ea8e5096eeb9cb1ef2bf37b5)
  - [Use webview_id and WebViewId in webview-related code](https://github.com/servo/servo/pull/31417/commits/3d869148cd759ec726268a4844bef5628ad98876)
  - [Improve logging of webview-related embedder events](https://github.com/servo/servo/pull/31417/commits/5801fbe270de836593f2c0acf4d7cdc3668964ef)
- [Don’t do anything in response to WebViewPaintingOrder](https://github.com/servo/servo/pull/31417/commits/2ce720dfdf79a7b93016578624d8c0d26320ba15)
  - [Remove WebViewPaintingOrder, since its payload is unreliable](https://github.com/servo/servo/pull/31417/commits/a3dbd6b4fc861cb82abf1375aff82c30f386a576)

delan committed:
Merge from main

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

mrobinson commented on components/compositing/compositor.rs:
```diff
@@ -1143,11 +1296,35 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
 
-    fn create_pipeline_details_for_frame_tree(&mut self, frame_tree: &SendableFrameTree) {
-        self.pipeline_details(frame_tree.pipeline.id).pipeline = Some(frame_tree.pipeline.clone());
+    fn update_pipeline_details_tree(
```
This has the sneaky side-effect of creating `PipelineDetails` if they don't exist, so I think the method name is really helpful to reveal this subtlety.

```suggestion
    fn create_or_update_pipeline_details_with_frame_tree(
```

delan replied:
done in b2528dc1bbe43ea18822c51cc3c2d6c10222bc0e

mrobinson commented on components/compositing/compositor.rs:
```diff
@@ -1143,11 +1296,35 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
+        }
+    }
+
+    fn remove_pipeline_details_tree(&mut self, pipeline_id: PipelineId) {
```
```suggestion
    fn remove_pipeline_details_recursively(&mut self, pipeline_id: PipelineId) {
```

delan replied:
done in done in b2528dc1bbe43ea18822c51cc3c2d6c10222bc0e

delan replied:
done in b2528dc1bbe43ea18822c51cc3c2d6c10222bc0e

mrobinson commented on components/compositing/compositor.rs:
```diff
@@ -1028,16 +1055,21 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
+    /// Set the root pipeline for our WebRender scene to a display list that consists of an iframe
+    /// for each visible top-level browsing context, applying a transformation on the root for
+    /// pinch zoom, page zoom, and HiDPI scaling.
+    fn update_root_pipeline_in_transaction(&self, transaction: &mut Transaction) {
```
I think this could be a bit more descriptive to make it obvious what this method is doing. How do you feel about:

```suggestion
    fn send_root_pipeline_display_list_in_transaction(&self, transaction: &mut Transaction) {
```

delan replied:
done in b2528dc1bbe43ea18822c51cc3c2d6c10222bc0e

mrobinson reviewed (approved):
This seems pretty good to me. I have a few minor nits that can be addressed before landing:

delan committed:
On MoveResizeWebView, only update root pipeline if rect changed

delan commented on components/compositing/compositor.rs:
```diff
@@ -1093,29 +1125,150 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
+            );
+        }
+
+        self.update_root_pipeline();
```
not quite, because it may still have moved, but we can check for that :)

delan replied:
done in be552012868c041679675130c550e7cdc85a7a5e

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
Rename IOCompositor methods for clarity

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan commented:
<https://github.com/servo/servo/actions/runs/8448204227/job/23139834020> shows an assertion failure in compositing WebViewManager::show, but so far I’ve been unable to reproduce it locally.

```
$ xvfb-run python3 ./mach smoketest --release
assertion failed: self.webviews.contains_key(&webview_id) (thread main, at /home/runner/work/servo/servo/components/compositing/webview.rs:61)
   0: servoshell::backtrace::print
   1: servoshell::main::{{closure}}
   2: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/alloc/src/boxed.rs:2021:9
      std::panicking::rust_panic_with_hook
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs:735:13
   3: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs:601:13
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/sys_common/backtrace.rs:170:18
   5: rust_begin_unwind
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs:597:5
   6: core::panicking::panic_fmt
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/panicking.rs:72:14
   7: core::panicking::panic
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/panicking.rs:127:5
   8: compositing::webview::WebViewManager<WebView>::show
   9: compositing::compositor::IOCompositor<Window>::show_webview
  10: servo::Servo<Window>::handle_events
  11: servoshell::app::App::handle_events
  12: servoshell::events_loop::EventsLoop::run_forever::{{closure}}
  13: winit::platform_impl::platform::x11::EventLoop<T>::run_return::single_iteration
  14: winit::platform_impl::platform::x11::EventLoop<T>::run
  15: winit::platform_impl::platform::EventLoop<T>::run
  16: servoshell::events_loop::EventsLoop::run_forever
  17: servoshell::app::App::run
  18: servoshell::main
  19: std::sys_common::backtrace::__rust_begin_short_backtrace
  20: std::rt::lang_start::{{closure}}
  21: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/ops/function.rs:284:13
      std::panicking::try::do_call
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs:504:40
      std::panicking::try
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs:468:19
      std::panic::catch_unwind
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panic.rs:142:14
      std::rt::lang_start_internal::{{closure}}
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/rt.rs:148:48
      std::panicking::try::do_call
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs:504:40
      std::panicking::try
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panicking.rs:468:19
      std::panic::catch_unwind
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/panic.rs:142:14
      std::rt::lang_start_internal
             at /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/std/src/rt.rs:148:20
  22: main
  23: <unknown>
  24: __libc_start_main
  25: _start
mozilla::detail::MutexImpl::~MutexImpl: pthread_mutex_destroy failed: Device or resource busy
Caught signal 11thread panicked while processing panic. aborting.
Redirecting call to abort() to mozalloc_abort

Servo was terminated by signal 11
```

delan commented:
<https://github.com/servo/servo/actions/runs/8449012673/job/23142218900> failed also, which is 2/3 of the attempts so far, so this is really something I’ll need to fix before landing.

delan commented:
I suspect the intermittent failures are because of a race that starts when the constellation handles NewWebView:

| (a) constellation WebViewOpened> embedder | (b) new_pipeline + add_pending_change |
| --- | --- |
| 1. servoshell &lt;WebViewOpened constellation | 1. constellation &lt;ActivateDocument script |
| 2. servoshell FocusWebView> constellation | 2. constellation CreateOrUpdateWebView> compositor |
| 3. servoshell &lt;WebViewFocused constellation |
| 4. servoshell ShowWebView> compositor |

(a) starts before (b), but I think they are otherwise independent, and if the compositor sees (a)4 before (b)2, we crash.

At first I thought that calling ShowWebView without ipc was the culprit, and that going through the constellation would help. Maybe it would, I haven’t tried yet. But I’m pretty sure that’s not enough, because the constellation can still receive ShowWebView on the compositor (and embedder) channel before it receives ActivateDocument on the script channel, which means it can still forward ShowWebView to the compositor before sending CreateOrUpdateWebView.

delan committed:
compositor: add event tracing; log webview ops even without ipc

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
Add temporary debug logging

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
Add more temporary debug logging

delan committed:
Remove temporary logging in compositor

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

wusyong commented:
smoke test is running `tests/html/close-on-load.html`. Maybe it's called way too soon before other initialization in the runner?

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
Remove temporary debug logging

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
Add temporary debug logging, but defer I/O until panic

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan commented:
Reproduced in more detail:

- pass https://github.com/wusyong/servo/actions/runs/8533456091/job/23376219529
- fail https://github.com/wusyong/servo/actions/runs/8533478848/job/23376274465

```
CrashLog([(Instant { tv_sec: 1396, tv_nsec: 738662526 }, Add(TopLevel(0,1))), (Instant { tv_sec: 1396, tv_nsec: 738664760 }, Show(TopLevel(0,1))), (Instant { tv_sec: 1397, tv_nsec: 798459860 }, HideAll), (Instant { tv_sec: 1397, tv_nsec: 798460001 }, RaiseToTop(TopLevel(0,1))), (Instant { tv_sec: 1397, tv_nsec: 836144678 }, Remove(TopLevel(0,1))), (Instant { tv_sec: 1397, tv_nsec: 836168353 }, HideAll), (Instant { tv_sec: 1397, tv_nsec: 836168422 }, Show(TopLevel(0,1)))]) (thread main, at /home/runner/work/servo/servo/components/compositing/webview.rs:100)
```

- (Instant { tv_sec: 1396, tv_nsec: 738662526 }, Add(TopLevel(0,1)))
- (Instant { tv_sec: 1396, tv_nsec: 738664760 }, Show(TopLevel(0,1)))
- (Instant { tv_sec: 1397, tv_nsec: 798459860 }, HideAll)
- (Instant { tv_sec: 1397, tv_nsec: 798460001 }, RaiseToTop(TopLevel(0,1)))
- (Instant { tv_sec: 1397, tv_nsec: 836144678 }, Remove(TopLevel(0,1)))
- (Instant { tv_sec: 1397, tv_nsec: 836168353 }, HideAll)
- (Instant { tv_sec: 1397, tv_nsec: 836168422 }, Show(TopLevel(0,1)))

delan committed:
Merge from main

delan committed:
Capture a backtrace with each crash log entry

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan commented:
fail <https://github.com/wusyong/servo/actions/runs/8534437489/job/23378836637>

1. libservo creates an IOCompositor, which adds and shows webview (0,1)
2. servoshell handles WebViewOpened by
    - sending the constellation a FocusWebView((0,1))
    - sending the compositor a RaiseWebViewToTop((0,1), true)
4. compositor handles RaiseWebViewToTop by hiding all webviews and raising webview (0,1) to top
5. compositor receives RemoveWebView from constellation and handles it by removing webview (0,1)
6. servoshell handles WebViewFocused by sending the compositor a ShowWebView((0,1), true)
7. compositor handles ShowWebView by hiding all webviews and showing webview (0,1)
8. compositor panics

```
CrashLog([
CrashLogEntry { instant: Instant { tv_sec: 1598, tv_nsec: 935719235 }, operation: Add(TopLevel(0,1)), backtrace: Backtrace [
{ fn: "compositing::webview::CrashLog::push" }, 
{ fn: "compositing::webview::WebViewManager<WebView>::add" }, 
{ fn: "compositing::compositor::IOCompositor<Window>::create" }, 
{ fn: "servo::Servo<Window>::new" }, 
{ fn: "servoshell::events_loop::EventsLoop::run_forever::{{closure}}" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::run_return::single_iteration" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::run" }, 
{ fn: "winit::platform_impl::platform::EventLoop<T>::run" }, 
{ fn: "servoshell::events_loop::EventsLoop::run_forever" }, 
{ fn: "servoshell::app::App::run" }, 
{ fn: "servoshell::main" }] }, 
CrashLogEntry { instant: Instant { tv_sec: 1598, tv_nsec: 937471407 }, operation: Show(TopLevel(0,1)), backtrace: Backtrace [
{ fn: "compositing::webview::CrashLog::push" }, 
{ fn: "compositing::compositor::IOCompositor<Window>::create" }, 
{ fn: "servo::Servo<Window>::new" }, 
{ fn: "servoshell::events_loop::EventsLoop::run_forever::{{closure}}" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::run_return::single_iteration" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::run" }, 
{ fn: "winit::platform_impl::platform::EventLoop<T>::run" }, 
{ fn: "servoshell::events_loop::EventsLoop::run_forever" }, 
{ fn: "servoshell::app::App::run" }, 
{ fn: "servoshell::main" }] }, 
CrashLogEntry { instant: Instant { tv_sec: 1598, tv_nsec: 938809020 }, operation: HideAll, backtrace: Backtrace [
{ fn: "compositing::webview::CrashLog::push" }, 
{ fn: "compositing::compositor::IOCompositor<Window>::raise_webview_to_top" }, 
{ fn: "servo::Servo<Window>::handle_events" }, 
{ fn: "servoshell::app::App::handle_events" }, 
{ fn: "servoshell::events_loop::EventsLoop::run_forever::{{closure}}" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::run_return::single_iteration" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::run" }, 
{ fn: "winit::platform_impl::platform::EventLoop<T>::run" }, 
{ fn: "servoshell::events_loop::EventsLoop::run_forever" }, 
{ fn: "servoshell::app::App::run" }, 
{ fn: "servoshell::main" }] }, 
CrashLogEntry { instant: Instant { tv_sec: 1598, tv_nsec: 939061422 }, operation: RaiseToTop(TopLevel(0,1)), backtrace: Backtrace [
{ fn: "compositing::webview::CrashLog::push" }, 
{ fn: "compositing::webview::WebViewManager<WebView>::raise_to_top" }, 
{ fn: "compositing::compositor::IOCompositor<Window>::raise_webview_to_top" }, 
{ fn: "servo::Servo<Window>::handle_events" }, 
{ fn: "servoshell::app::App::handle_events" }, 
{ fn: "servoshell::events_loop::EventsLoop::run_forever::{{closure}}" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::run_return::single_iteration" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::run" }, 
{ fn: "winit::platform_impl::platform::EventLoop<T>::run" }, 
{ fn: "servoshell::events_loop::EventsLoop::run_forever" }, 
{ fn: "servoshell::app::App::run" }, 
{ fn: "servoshell::main" }] }, 
CrashLogEntry { instant: Instant { tv_sec: 1599, tv_nsec: 835868951 }, operation: Remove(TopLevel(0,1)), backtrace: Backtrace [
{ fn: "compositing::webview::CrashLog::push" }, 
{ fn: "compositing::webview::WebViewManager<WebView>::remove" }, 
{ fn: "compositing::compositor::IOCompositor<Window>::receive_messages" }, 
{ fn: "servo::Servo<Window>::handle_events" }, 
{ fn: "servoshell::app::App::handle_events" }, 
{ fn: "servoshell::events_loop::EventsLoop::run_forever::{{closure}}" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::drain_events::{{closure}}::{{closure}}" }, 
{ fn: "winit::platform_impl::platform::x11::event_processor::EventProcessor<T>::process_event" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::run_return::single_iteration" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::run" }, 
{ fn: "winit::platform_impl::platform::EventLoop<T>::run" }, 
{ fn: "servoshell::events_loop::EventsLoop::run_forever" }, 
{ fn: "servoshell::app::App::run" }, 
{ fn: "servoshell::main" }] }, 
CrashLogEntry { instant: Instant { tv_sec: 1599, tv_nsec: 836099380 }, operation: HideAll, backtrace: Backtrace [
{ fn: "compositing::webview::CrashLog::push" }, 
{ fn: "compositing::compositor::IOCompositor<Window>::show_webview" }, 
{ fn: "servo::Servo<Window>::handle_events" }, 
{ fn: "servoshell::app::App::handle_events" }, 
{ fn: "servoshell::events_loop::EventsLoop::run_forever::{{closure}}" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::drain_events::{{closure}}::{{closure}}" }, 
{ fn: "winit::platform_impl::platform::x11::event_processor::EventProcessor<T>::process_event" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::run_return::single_iteration" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::run" }, 
{ fn: "winit::platform_impl::platform::EventLoop<T>::run" }, 
{ fn: "servoshell::events_loop::EventsLoop::run_forever" }, 
{ fn: "servoshell::app::App::run" }, 
{ fn: "servoshell::main" }] }, 
CrashLogEntry { instant: Instant { tv_sec: 1599, tv_nsec: 836201541 }, operation: Show(TopLevel(0,1)), backtrace: Backtrace [
{ fn: "compositing::webview::CrashLog::push" }, 
{ fn: "compositing::compositor::IOCompositor<Window>::show_webview" }, 
{ fn: "servo::Servo<Window>::handle_events" }, 
{ fn: "servoshell::app::App::handle_events" }, 
{ fn: "servoshell::events_loop::EventsLoop::run_forever::{{closure}}" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::drain_events::{{closure}}::{{closure}}" }, 
{ fn: "winit::platform_impl::platform::x11::event_processor::EventProcessor<T>::process_event" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::run_return::single_iteration" }, 
{ fn: "winit::platform_impl::platform::x11::EventLoop<T>::run" }, 
{ fn: "winit::platform_impl::platform::EventLoop<T>::run" }, 
{ fn: "servoshell::events_loop::EventsLoop::run_forever" }, 
{ fn: "servoshell::app::App::run" }, 
{ fn: "servoshell::main" }] }]) (thread main, at /home/runner/work/servo/servo/components/compositing/webview.rs:115)
```

delan committed:
Proper error handling without panicking in WebViewManager

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
Clean up imports in constellation

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

wusyong commented:
@delan Thanks for addressing the rest of the issues in the PR! I was absent last week, but I'll catch up in the next few days.
And hopefully try to see if I can work on multiple views in same window from my project as well.
